[MM-24714]  Render markdown in email notifications

Zeezee1210 opened the pull request:
#### Summary
This PR adds a function in the utils/markdown.go to take a string containing markdown and then return the string converted to HTML tags. It's then called when creating the body of notification emails to convert the post message markdown to HTML.

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/17613

#### Release Note

```release-note
Added rendering for posts containing markdown in email notifications
```

mattermod commented:
Hello @Zeezee1210,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

hmhealey reviewed (commented):
Thanks for the PR, @Zeezee1210! This is something we've wanted for a long time, but I honestly didn't realize the had a proper Markdown library for use on the server now.

This looks good to me, but I'm going to defer my review to @agnivade since he's worked with the server's Markdown code much more recently than I have.

agnivade reviewed (commented):
Thank you @Zeezee1210 for the PR!

Zeezee1210 commented on app/notification_email_test.go:
```diff
@@ -825,3 +825,96 @@ func TestLandingLinkPermalink(t *testing.T) {
 	require.Contains(t, body, teamURL+"/pl/"+post.Id, fmt.Sprintf("Expected email text '%s'. Got %s", teamURL, body))
 }
+
+func TestMarkdownLinksRender(t *testing.T) {
```
Hi @agnivade,
Thanks a lot for the great review and appreciate your feedback.
I've switched this to table-driven test (which I've hopefully done correctly)...when it comes to adding more tests I popped in a few based on looking at the tests used in the markdown_test.go but if there's any more specific ones you'd want adding just say.

Zeezee1210 commented on app/notification_email_test.go:
```diff
@@ -825,3 +825,96 @@ func TestLandingLinkPermalink(t *testing.T) {
+	body, err := th.App.getNotificationEmailBody(recipient, post, channel, channelName, senderName, teamName, teamURL, emailNotificationContentsType, true, translateFunc, "user-avatar.png")
+	expect := "This is <a href=\"https://mattermost.com\">Mattermost</a>"
+	require.NoError(t, err)
+	require.Contains(t, body, expect, fmt.Sprintf("Expected email text '%s'. Got %s", expect, body))
```
Yes, I based the tests on the previous ones. I've removed the string as I don't see a need when testify is good enough.

Zeezee1210 commented on app/notification_email_test.go:
```diff
@@ -825,3 +825,96 @@ func TestLandingLinkPermalink(t *testing.T) {
+	teamName := "testteam"
+	teamURL := "http://localhost:8065/landing#/testteam"
+	emailNotificationContentsType := model.EmailNotificationContentsFull
+	translateFunc := i18n.GetUserTranslations("en")
```
Agreed, cleaner just passing directly for a test so I've changed that.

mkraft commented:
@Zeezee1210 This is going to be really awesome!

agnivade commented on app/notification_email_test.go:
```diff
@@ -825,3 +825,99 @@ func TestLandingLinkPermalink(t *testing.T) {
+				Message: tt.args,
+			}
+			got, err := th.App.getNotificationEmailBody(recipient, post, channel, "ChannelName", "sender", "testteam", "http://localhost:8065/landing#/testteam", model.EmailNotificationContentsFull, true, i18n.GetUserTranslations("en"), "user-avatar.png")
+			if err != nil {
+				t.Fatalf("error: %v", err)
+			}
```
Change this to `require.NoError(t, err)`

Zeezee1210 replied:
haha damn! I did have this but for some reason went with the t.Fatalf.

agnivade commented on app/notification_email_test.go:
```diff
@@ -825,3 +825,99 @@ func TestLandingLinkPermalink(t *testing.T) {
+			if err != nil {
+				t.Fatalf("error: %v", err)
+			}
+			require.Contains(t, got, tt.want)
```
Should it be `Contains` or `Equal`? I'd assume it should be equal.

Zeezee1210 replied:
I believe this should be `Contains` as it is testing against a render of the whole email HTML and we're only looking for the message portion.

agnivade replied:
Ah yes, I missed that the entire email contains lots of other stuff.

agnivade reviewed (commented):
Nice work! Just a couple of finishing touches needed.

agnivade reviewed (approved):
Thank you @Zeezee1210 ! 

I think I somehow missed that we already have a custom markdown renderer in our codebase. I am wondering how much of a parity we have with that. Would it be possible for you to try the `shared/markdown.RenderHTML` function? 

I'd also want to move these two functions from inside utils to the `shared/markdown` package.

Zeezee1210 commented:
@agnivade, I did spot that when I first started working on it and gave it a go. Unfortunately it only worked for things like links, I initially switched to the current package to support emphasis. I suspect it won't have supported strikethrough/tables either.

agnivade commented:
Sounds good, thanks for the thorough investigation! @hmhealey - I think we should unify the two modules at some point. If the goldmark parser is a superset, we should be able to replace some of the home-grown code.

hmhealey commented:
Yeah, I agree that it would make sense to combine them. Apart from additions like at-mentions, we do have a few places where we deviate from proper Markdown that we'd probably like to address, but we aren't 100% consistent even between mobile and web yet, so those would probably be acceptable for the time being.

For context around the `shared/markdown` package, that was added fairly early on to do some URL rewriting back when the image proxy was first added, and I think it's also used more recently for mention parsing and pulling out URLs for file previews. It only has the ability to parse things that will affect those (essentially just links and code blocks).

agnivade commented:
The goldmark parser is fully extendable and allows for callbacks when walking the entire markdown tree. So extracting URLs, mention parsing should all be doable.

furqanmlk reviewed (commented):
Thanks @Zeezee1210 

I have verified following formate, all looks good except "**Block Quotes**"
**Text Style** ✅ 
**Links** ✅ 
**Headings** ✅ 
**Lists** ✅ 
**Code Block** ✅ 
**Block quotes** 🔴 
**Tables** ✅ 


![image](https://user-images.githubusercontent.com/6909890/130137091-c1c8c1d5-e6a9-4805-a745-6198f1f3f31d.png)

Zeezee1210 commented:
Thanks for the testing @furqanmlk, I took a look at the blockquote issue you spotted. It appears there was a call to html.EscapeString just prior to the markdown conversion and therefore the > was escaped and didn't get recognised as start of blockquote markdown. 

I've removed that call as I don't believe it was needed, and also added a test for blockquotes in the table tests. Looks to have sorted it on my local dev build.

hmhealey reviewed (changes requested):
That sounds like it should be fine to remove the `html.EscapeString` call since I assume the Markdown library does its own escaping, but could you add a test to confirm that? Something as simple as the input being `<b>not bold</b>` would be fine

Zeezee1210 commented:
@hmhealey good point. I might have to re-think it. 
Adding that test to current code shows it converts `<b>not bold</bold>` to `<!-- raw HTML omitted -->not bold<!-- raw HTML omitted -->`

furqanmlk commented:
> Thanks for the testing @furqanmlk, I took a look at the blockquote issue you spotted. It appears there was a call to html.EscapeString just prior to the markdown conversion and therefore the > was escaped and didn't get recognised as start of blockquote markdown.
> 
> I've removed that call as I don't believe it was needed, and also added a test for blockquotes in the table tests. Looks to have sorted it on my local dev build.

Thanks @Zeezee1210 
I re-tested, I don't see quote in correct quote format, is that expected. ?

![image](https://user-images.githubusercontent.com/6909890/130275274-5ae3707f-bfcd-4168-a9e4-bf554017d1f5.png)

Zeezee1210 commented:
> Thanks @Zeezee1210 
> 
> I re-tested, I don't see quote in correct quote format, is that expected. ?
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/6909890/130275274-5ae3707f-bfcd-4168-a9e4-bf554017d1f5.png)
> 
> 

Hi @furqanmlk,
I would expect a markdown blockquote to render like this:
![image](https://user-images.githubusercontent.com/14242437/130276057-7c65024c-0870-4079-96b2-e15bf20db536.jpeg)

With a bar on the left and the text indented. Further reference here:
https://www.w3schools.io/file/markdown-blockquotes/

furqanmlk commented:
> > Thanks @Zeezee1210
> > I re-tested, I don't see quote in correct quote format, is that expected. ?
> > ![image](https://user-images.githubusercontent.com/6909890/130275274-5ae3707f-bfcd-4168-a9e4-bf554017d1f5.png)
> 
> Hi @furqanmlk,
> I would expect a markdown blockquote to render like this:
> ![image](https://user-images.githubusercontent.com/14242437/130276057-7c65024c-0870-4079-96b2-e15bf20db536.jpeg)
> 
> With a bar on the left and the text indented. Further reference here:
> https://www.w3schools.io/file/markdown-blockquotes/

Yeah, I don't see that vertical  bar in my email

![image](https://user-images.githubusercontent.com/6909890/130276350-2c940af3-7b68-40f9-81b3-768283e7a2d8.png)

Zeezee1210 commented:
> > > Thanks @Zeezee1210
> 
> > > I re-tested, I don't see quote in correct quote format, is that expected. ?
> 
> > > ![image](https://user-images.githubusercontent.com/6909890/130275274-5ae3707f-bfcd-4168-a9e4-bf554017d1f5.png)
> 
> > 
> 
> > Hi @furqanmlk,
> 
> > I would expect a markdown blockquote to render like this:
> 
> > ![image](https://user-images.githubusercontent.com/14242437/130276057-7c65024c-0870-4079-96b2-e15bf20db536.jpeg)
> 
> > 
> 
> > With a bar on the left and the text indented. Further reference here:
> 
> > https://www.w3schools.io/file/markdown-blockquotes/
> 
> 
> 
> Yeah, I don't see that vertical  bar in my email
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/6909890/130276350-2c940af3-7b68-40f9-81b3-768283e7a2d8.png)
> 
> 

Hmm odd. Does the raw html of the email still have the correct html tags? Something probably like this:
```
<blockquote>
<p>block quotes</p>
</blockquote>
```

furqanmlk commented:
> > > > Thanks @Zeezee1210
> > 
> > 
> > > > I re-tested, I don't see quote in correct quote format, is that expected. ?
> > 
> > 
> > > > ![image](https://user-images.githubusercontent.com/6909890/130275274-5ae3707f-bfcd-4168-a9e4-bf554017d1f5.png)
> > 
> > 
> > > 
> > 
> > 
> > > Hi @furqanmlk,
> > 
> > 
> > > I would expect a markdown blockquote to render like this:
> > 
> > 
> > > ![image](https://user-images.githubusercontent.com/14242437/130276057-7c65024c-0870-4079-96b2-e15bf20db536.jpeg)
> > 
> > 
> > > 
> > 
> > 
> > > With a bar on the left and the text indented. Further reference here:
> > 
> > 
> > > https://www.w3schools.io/file/markdown-blockquotes/
> > 
> > 
> > Yeah, I don't see that vertical  bar in my email
> > ![image](https://user-images.githubusercontent.com/6909890/130276350-2c940af3-7b68-40f9-81b3-768283e7a2d8.png)
> 
> Hmm odd. Does the raw html of the email still have the correct html tags? Something probably like this:
> 
> ```
> <blockquote>
> <p>block quotes</p>
> </blockquote>
> ```

Looks like, it's there

![image](https://user-images.githubusercontent.com/6909890/130277120-55d61794-9066-452a-8dd3-b4ab3c6cecd3.png)

Zeezee1210 commented:
> > > > > Thanks @Zeezee1210
> 
> > > 
> 
> > > 
> 
> > > > > I re-tested, I don't see quote in correct quote format, is that expected. ?
> 
> > > 
> 
> > > 
> 
> > > > > ![image](https://user-images.githubusercontent.com/6909890/130275274-5ae3707f-bfcd-4168-a9e4-bf554017d1f5.png)
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > Hi @furqanmlk,
> 
> > > 
> 
> > > 
> 
> > > > I would expect a markdown blockquote to render like this:
> 
> > > 
> 
> > > 
> 
> > > > ![image](https://user-images.githubusercontent.com/14242437/130276057-7c65024c-0870-4079-96b2-e15bf20db536.jpeg)
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > With a bar on the left and the text indented. Further reference here:
> 
> > > 
> 
> > > 
> 
> > > > https://www.w3schools.io/file/markdown-blockquotes/
> 
> > > 
> 
> > > 
> 
> > > Yeah, I don't see that vertical  bar in my email
> 
> > > ![image](https://user-images.githubusercontent.com/6909890/130276350-2c940af3-7b68-40f9-81b3-768283e7a2d8.png)
> 
> > 
> 
> > Hmm odd. Does the raw html of the email still have the correct html tags? Something probably like this:
> 
> > 
> 
> > ```
> 
> > <blockquote>
> 
> > <p>block quotes</p>
> 
> > </blockquote>
> 
> > ```
> 
> 
> 
> Looks like, it's there
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/6909890/130277120-55d61794-9066-452a-8dd3-b4ab3c6cecd3.png)
> 
> 

That looks like a difference in mail client html rendering then. I have a test email in ProtonMail client which shows the bar, but on Apple iOS mail app it chooses not to show the bar. Appears to be the correct html tagging though.

Zeezee1210 commented:
> That sounds like it should be fine to remove the `html.EscapeString` call since I assume the Markdown library does its own escaping, but could you add a test to confirm that? Something as simple as the input being `<b>not bold</b>` would be fine

Hi @hmhealey,
It would be great if you could take a look at the latest commit to see what you think.  The issue I found was using html.EscapeString on the entire postmessage first meant that the markdown parser missed the blockquote angle bracket. Then when I removed that step, you rightly showed that it would then affect html tags in the post message. 
So I added a cleaning step in the markdown function to escape strings in matching pairs of html tags and leave the others. Anything not caught should be protected by the markdown package that doesn't allow raw html by replacing it with ` <!-- raw HTML omitted -->`

hmhealey commented on utils/markdown.go:
```diff
@@ -35,12 +37,19 @@ func StripMarkdown(markdown string) (string, error) {
 
 // MarkdownToHTML takes a string containing Markdown and returns a string with HTML tagged version
 func MarkdownToHTML(markdown string) (string, error) {
+	re := regexp.MustCompile(`<(.*)>.*?|<(.*) />`)
```
This regex isn't going to work unfortunately. Obviously we can't use `html.EscapeString` on the whole thing because of the issue with block quotes, but I'd be worried we'd miss something with anything like this that's much more complicated than just replacing the unsafe characters with HTML entities. Generally, trying to use regexes to do much with HTML is a bad idea.

Zeezee1210 replied:
Hi @hmhealey, how about doing the opposite? 
I've re-added the html.EscapeString step but within utils.MarkdownToHTML I now search for > at the beginning of a line to then unescape just that for blockquote parsing?

hmhealey reviewed (changes requested):
That's unfortunate that there's no built-in option in Goldmark to just escape raw HTML and render it as plain text. Is that maybe something we could override by messing with the renderer?

Zeezee1210 commented:
> That's unfortunate that there's no built-in option in Goldmark to just escape raw HTML and render it as plain text. Is that maybe something we could override by messing with the renderer?

From what I have read and understand there is a renderer option to allow raw html (called unsafe) which would then pass back a message post with `<b>bold</b>` but then obviously as that gets put in an email it's rendered as html because it hasn't encoded it.

The reason I put the regexp in was it needs to be just good enough to not encode the blockquote angle bracket but capture most html tagging and encode it. If it doesn't catch it, then the default behaviour of goldmark is to not allow unsafe html and strips it entirely. Which makes the regexp safer because there's the second sweep looking for html and removes it.

If you think that's still not a good idea, I'll keep having a think about it...would also appreciate any suggestions ;)

Zeezee1210 commented:
> > That's unfortunate that there's no built-in option in Goldmark to just escape raw HTML and render it as plain text. Is that maybe something we could override by messing with the renderer?
> 
> From what I have read and understand there is a renderer option to allow raw html (called unsafe) which would then pass back a message post with `<b>bold</b>` but then obviously as that gets put in an email it's rendered as html because it hasn't encoded it.
> 
> The reason I put the regexp in was it needs to be just good enough to not encode the blockquote angle bracket but capture most html tagging and encode it. If it doesn't catch it, then the default behaviour of goldmark is to not allow unsafe html and strips it entirely. Which makes the regexp safer because there's the second sweep looking for html and removes it.
> 
> If you think that's still not a good idea, I'll keep having a think about it...would also appreciate any suggestions ;)

@hmhealey, having re-read this and just in case I didn't explain it clearly. The regexp is acting as an "allowlist" and anything it captures will then be encoded and show safely in a post message...everything else will be removed by the Goldmark renderer.

hmhealey commented:
Apologies for the delay. I ended up being out all of last week, and I forgot to check in on this PR.

I think I understand what you were trying to do before, but I was worried there might have been ways to either get around the regex or cause it to have performance problems (which have come up a lot in the web app's Markdown parser in the past).

That said, I actually really like the new way of doing this where we start with everything escaped and then unescape what we need for block quotes. Starting with it escaped makes me comfortable that we shouldn't let anything through by mistake, and the regex is simple enough that it can't cause any performance problems there. I wasn't sure how we'd be able to fix that without doing something more complicated like forking the Markdown library. Nice job!

furqanmlk commented:
@Zeezee1210 Are we ok without `Quote` format ?
I still don't vertical bar with quote in the email.

### Mattermost
![image](https://user-images.githubusercontent.com/6909890/131406999-345a6d17-0185-4a06-b445-948076d9fac1.png)

### Email
![image](https://user-images.githubusercontent.com/6909890/131406943-24ffdef9-78d6-4b7a-a73c-8cc995e746b0.png)

Zeezee1210 commented:
> @Zeezee1210 Are we ok without `Quote` format ?
> 
> I still don't vertical bar with quote in the email.
> 
> 
> 
> ### Mattermost
> 
> ![image](https://user-images.githubusercontent.com/6909890/131406999-345a6d17-0185-4a06-b445-948076d9fac1.png)
> 
> 
> 
> ### Email
> 
> ![image](https://user-images.githubusercontent.com/6909890/131406943-24ffdef9-78d6-4b7a-a73c-8cc995e746b0.png)
> 
> 

Hi @furqanmlk,
I'm ok with that as the HTML is correct and the rendering is client specific/dependent so can't control it.

mattermod commented:
Will try to auto merge this PR once all tests and checks are passing. This might take up to an hour.

agnivade commented:
/update-branch

agnivade commented:
/update-branch

mattermod commented:
Trying to auto merge this PR.

mattermod commented:
Pull Request successfully merged
SHA: 0b79defcbb8c0b927fd4610629762e9153ae54a9
